Chore/test deploy configs

MisterTimn opened the pull request:
#### 📁 Related issues
#354 

#### ✍️ Description

A bash script has been added `/test/deploy/validate-configs.sh`
This script will test run CSS with all the different default configs by looping over `config/*.json`
For https (`example-https-file.json` and `https-cli-file.json`) certs are set (through cli params or config file overwrites) and base URL is changed.

This script is run in the `test-configs` job. The job has a `tenforce/virtuoso` service for sparql config validation and also generates a self-signed cert/key for https. All configs get tested, in case of failure the logs from CSS get printed and the script moves on to the next config. At the end an overview printout is provided:

![image](https://user-images.githubusercontent.com/4209558/164002464-2199ff58-c32e-47a2-96fb-cbe1078d0333.png)

If one of the tests has failed, the overall job will fail.

## Improvements
### Conditional command changes
Additional exceptions like in the case of https can be set up but this might become cumbersome in the future as configs get added or changed. Perhaps we might move to a solution where default configs are accompanied with an entry in a separate validation config file that provides a string to append to the `community-solid-server` command. With the addition #1264 and dynamic CLI parameters, we can also enforce all future configs to have parameters that allow it to be run without editing config files (like in the case of `example-https-file.json`).

### Validation / deployment testing / config testing?
Right now, the job name is `test-configs` and the script is called `validate-configs`, I'm not sure how we best name this step and where we position it in the pipeline. It's more a form of deployment testing with different configs than a true test... 

### ✅ PR check list

Before this pull request can be merged, a core maintainer will check whether
  * [x] this PR is labeled with the correct semver label
    - semver.patch: Backwards compatible bug fixes.
    - semver.minor: Backwards compatible feature additions.
    - semver.major: Breaking changes. This includes changing interfaces or configuration behaviour.
  * [x] the correct branch is targeted. Patch updates can target main, other changes should target the latest versions/* branch.
  * [ ] the RELEASE_NOTES.md document in case of relevant feature or config changes.
  * [ ] any relevant documentation was updated to reflect the changes in this PR.

RubenVerborgh commented:
Great!

Is this the right target branch though? I see a couple of unrelated commits.

MisterTimn commented:
> Is this the right target branch though? I see a couple of unrelated commits.

Oh, yea I rebased off the version/4.0.0 branch before opening the PR but then targeted main. I think this best targets main since its unrelated to the CSS code itself, maybe after the 4.0.0 release this can be cleaned up.

joachimvh commented on .github/workflows/ci.yml:
```diff
@@ -119,6 +119,29 @@ jobs:
+          node-version: '16.x'
+      - name: Check out repository
+        uses: actions/checkout@v3
+      - name: Generate certificates https config
```
```suggestion
      - name: Generate certificates for HTTPS configurations
```

joachimvh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env bash
```
I only have a vague knowledge of bash scripts so can't really comment on the contents. @rubensworks or @RubenVerborgh are better suited.

joachimvh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,88 @@
+  CSS_OPTS="-p 8888 -l warn -s http://localhost:4000/sparql"
+  CURL_STATEMENT='http://localhost:8888'
+  # HTTPS config needs a base-url override + keys and certs
+  if [[ $CONFIG_NAME =~ "https" ]]; then
+    sed -i -E "s/(\W+\"options_key\".*\").+(\".*)/\1server.key\2/" $CONFIG_PATH
+    sed -i -E "s/(\W+\"options_cert\".*\").+(\".*)/\1server.cert\2/" $CONFIG_PATH
+    CSS_OPTS="$CSS_OPTS -b https://localhost:8888 --httpsKey server.key --httpsCert server.cert"
+    CURL_STATEMENT='-k https://localhost:8888'
+  fi
```
I would have 2 separate if-statements for the 2 different HTTPS configs so only one of them gets the CLI params, and only one of them gets the changed config. You can also changes the paths in `example-https-file.json` should that make testing easier.

joachimvh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,88 @@
+echo "----------------------------------------"
+echo -e $SUCCESSES
+echo -e $FAILURES
+exit $VALIDATION_FAILURE
```
Missing EOL.

joachimvh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,88 @@
+npm install -g solid-community-server-*.tgz --loglevel warn
+rm solid-community-server-*.tgz
+
+run_server_with_config () {
```
Can we remove `validate-package.sh` and its corresponding CI check since this script does the same and more?

MisterTimn replied:
Should there be a script shortcut added for this test in package-lock.json as well then? Or just replace `npm run test:deploy` with this script?

RubenVerborgh replied:
I'd replace indeed. Or maybe this script can optionally take argument(s) to indicate which files to process.

MisterTimn replied:
That is certainly possible, might come in handy.
So either just `npm run test:deploy`, no arguments checks all configs, if argument is provided will test given config

Or we make two entries:
`npm run test:deploy` would be the version that takes the argument of a config file, no argument runs 'default.json'
`npm run test:deploy-configs` or something similar would then be all configs

RubenVerborgh replied:
One entry works for me. Then we can still `npm run test:deploy -- tests`.

MisterTimn replied:
Might be best in that case to split things up more, since in this script we do a `npm pack` followed by an `npm install`. Which would be overkill if you just want to validate config locally (just build and run in that case, which seems quicker)

RubenVerborgh replied:
We could—although it's really a deploy test so it's doing the right thing for me.

MisterTimn commented:
The script can now be handed one or multiple arguments, either full path, name, name.json or config/name (see example below).

```bash
$ npm run test:deploy default file.json config/example-https-file config/https-file-cli.json

> @solid/community-server@4.0.0 test:deploy
> test/deploy/validate-configs.sh "default" "file.json" "config/example-https-file" "config/https-file-cli.json"

Deployment testing 4 configs
 - config/default.json
 - config/file.json
 - config/example-https-file.json
 - config/https-file-cli.json
 ```
 
No arguments will start test for all configs.
```bash
$ npm run test:deploy

> @solid/community-server@4.0.0 test:deploy
> test/deploy/validate-configs.sh

Deployment testing all configs!
 - config/default.json
 - config/dynamic.json
 - config/example-https-file.json
 - config/file-no-setup.json
 - config/file.json
 - config/https-file-cli.json
 - config/memory-subdomains.json
 - config/path-routing.json
 - config/quota-file.json
 - config/restrict-idp.json
 - config/sparql-endpoint-no-setup.json
 - config/sparql-endpoint.json
 - config/sparql-file-storage.json
 ```
 
 Added `-f` argument to css run with a folder for automated cleanup of all files after tests, so it can be used locally.

 **!** When you run locally, sparql configs will fail unless you have an endpoint on `https://localhost:4000/sparql`

Certificate generation was moved inside the script (instead of from CI run) to allow local deploy:test of https (now it will get generated twice in case all configs are run, don't think this is an issue but might be better to generate at start script (not in the https conditional)

The example-https-file gets copied and placed back so original values are restored. Alternatively, we just change the paths in the file like @joachimvh suggested, but maybe best to keep it in a dir still (to indicate clearly it needs a path). Something like `ssl/server.key`, then either create/remove that folder everytime or leave it.

Perhaps I should look into writing some small Developer docs on what to do/check when new default configs get added?

MisterTimn commented:
Im seeing the unit tests on ubuntu fail because they are running `deploy:test` as well. Remove this step from all CI unit tests as we can depend on the separate deploy-configs job for this?

RubenVerborgh commented:
> Im seeing the unit tests on ubuntu fail because they are running `deploy:test` as well. Remove this step from all CI unit tests as we can depend on the separate deploy-configs job for this?

Absolutely, yes!

RubenVerborgh commented on .github/workflows/ci.yml:
```diff
@@ -119,6 +119,29 @@ jobs:
       - name: Run integration tests
         run: npm run test:integration
 
+  test-configs:
```
Shall we add passing this suite as a precondition to `docker-main-branch` and `docker-versions-branch` and `publish-docs`? Also we probably want some consistency in our `needs:`; I see variations.

RubenVerborgh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,118 @@
+
+TEST_NAME="Deployment testing"
+declare -a CONFIG_ARRAY
+if [[ $# -gt 0 ]]; then
```
Some comments perhaps regarding expected arguments?

RubenVerborgh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,118 @@
+# Script to validate the packaged module and configs
+
+# Ensure our workdir is that of the project root
+cd "${0%/*}/../.." || exit 
```
Exit with warning / error code?

RubenVerborgh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,118 @@
+fi
+printf " - %s\n" "${CONFIG_ARRAY[@]}"
+
+mkdir -p .tmp
```
Should be in `.gitignore`.

joachimvh replied:
Could also just use `/test/tmp` which already is in the `.gitignore` and put both folder (`.tmp` and `.data`) in there.

RubenVerborgh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,118 @@
+  CONFIG_PATH=$1
+  CONFIG_NAME=$2
+
+  mkdir -p .data
```
In `.gitignore`?

RubenVerborgh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,118 @@
+    if curl -sfkI -X GET --retry 15 --retry-connrefused --retry-delay 1 $CSS_BASE_URL > .tmp/"$CONFIG_NAME"-curl; then
+      echo "$TEST_NAME($CONFIG_NAME) - SUCCESS: server reached"
+      FAILURE=0
+    fi
+    if [ $FAILURE -eq 1 ]; then
```
Could just be an `else`?

RubenVerborgh replied:
Could just be an `else`?

RubenVerborgh reviewed (approved):
Super cool stuff!

joachimvh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,118 @@
+declare -a CONFIG_ARRAY
+if [[ $# -gt 0 ]]; then
+    for CONFIG in "$@"; do
+      CONFIG=$(ls {"$CONFIG",config/"$CONFIG",config/"$CONFIG".json,"$CONFIG".json} 2> /dev/null)
```
If I understand correctly this means I can call this script with `config/file.json`, `file.json`, `file` and `config/file` as valid input parameters all corresponding to `config/file.json`? Would only pick one of those as a valid option. Personally I would require the "full" path so `config/file.json` for least confusion.

MisterTimn replied:
Your assumption is correct. I'll change it back to only full path, might get confusing indeed.

joachimvh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,118 @@
+
+  # HTTPS config specifics: self-signed key/cert + CSS base URL override
+  if [[ $CONFIG_NAME =~ "https" ]]; then
+    openssl req -x509 -nodes -days 1 -newkey rsa:2048 -keyout .tmp/server.key -out .tmp/server.cert -subj '/CN=localhost' &>/dev/null
```
Didn't this already happen in the `ci.yml`?

MisterTimn replied:
Yes but I'm going to remove it there. If we only do it in ci.yml we can't test:deploy locally. Unless we keep self-signed certs in our repo for testing, which is also an option.

joachimvh replied:
Removing from the CI is fine, was just confused why I saw it twice.

joachimvh commented on test/deploy/validate-configs.sh:
```diff
@@ -0,0 +1,118 @@
+
+echo "$TEST_NAME - Cleanup"
+npm uninstall -g @solid/community-server
+mv .tmp/example-https-file.json config/example-https-file.json 2>/dev/null || true
```
Why is the temporary HTTPS config being moved to the default one?

MisterTimn replied:
Because I overwrite the certs options and want to place back the original file so they don't get committed after running this script locally.
But might as well just change the certs path and generate them in a directory, `ssl/server.cert` for example? This makes it clear we need a path. Then we just add `ssl/` to `.gitignore`.

joachimvh replied:
Ah I misread the script. I assumed `.tmp/example-https-file.json` was the one being run, but it's the backup. Would it be possible to run `.tmp/example-https-file.json` instead? That way we don't have to worry about accidentally changing the default config in case something goes wrong.

MisterTimn replied:
yea that makes a lot more sense then what I'm doing now ^^

MisterTimn commented:
Requested changes are implemented

- All files needed are generated/cleaned from `test/tmp` folder which is already in gitignore
- `test:deploy` only called in its separate job and `needs` updated
- fixed an issue with parameter expansion
- Now only takes full config path as optional parameter (relative from project directory or absolute)

The pipeline did fail again for unit tests because of `npm run test:deploy` still being called, but I removed this step from the `test-unit` job in https://github.com/CommunitySolidServer/CommunitySolidServer/pull/1268/commits/39732ac278ce6f3158999c340a50dc019f45a54c

RubenVerborgh commented:
So nice to see this pass. Great work! @joachimvh Let's make `test-configs` mandatory in GitHub after this is merged.

RubenVerborgh commented:
@joachimvh Do we have graceful shutdown instead of kill? Like a HUP? Or is this the safe way?

joachimvh commented:
> @joachimvh Do we have graceful shutdown instead of kill? Like a HUP? Or is this the safe way?

We don't. The only way to gracefully shutdown is if you start the server through code.
